rfcs/remote-schema-permissions.md

```diff
@@ -0,0 +1,244 @@
+```
+
+In short, any list reference can be replaced by `*` to mean "all valid values" for that list. For example,
+
+1. For an object type, like "User", `fields` can be `*`
+2. Even the root level definition can take `*` like `allowed_custom_scalars: *`
+
+
+## Role-based argument presets
+
+The above API only illustrates schema masking. We need to incorporate argument presets as well to provide "filtered" results. The API described in this section is the final proposed API.
+
+Each field in GraphQL can take arguments as well. Arguments must belong to input types i.e Scalar, Enum, or Input Object (or list of these).
+
+Role-based argument presets can basically inject values from session variables or static values during execution. The presetted argument should also get removed from the exposed role-based schema.
+
+The proposed API:
+
+```
+type: add_remote_schema_permissions
+args:
+  remote_schema: comm
+  role: user
+  definition:
+    allowed_objects:
+    - name: Query
+      fields: 
+      - name: hello
+      - name: user
+        argument_presets:
+        - name: "id"
+          value:
+            from_session_variable: "x-hasura-user-id"
+    - name: User
+      fields:
+      - name: a
+      - name: b
+      - name: c
+```
+
+Argument presets can also go arbitrarily deep if the input type is Input Object (this is very similar to how we define the mapping in remote relationships):
+
+```
+type: add_remote_schema_permissions
+args:
+  remote_schema: comm
+  role: user
+  definition:
+    allowed_objects:
+    - name: Query
+      fields:
+        - name: hello
+        - name: user
+          argument_presets:
+          - name: "where"
+            input_field:
+            - name: "id"
+              input_field:
+              - name: "_eq"
+                value:
+                  from_session_variable: "x-hasura-user-id"
+          - name: "limit"
+            value:
+              static: 1
+    - name: User
+      fields:
+      - name: a
+      - name: b
+      - name: c
+```
+
+There can be more than one argument preset per field:
+
+```
+type: add_remote_schema_permissions
+args:
+  remote_schema: comm
+  role: user
+  definition:
+    allowed_objects:
+    - name: Query
+      fields: 
+      - name: hello
+      - name: user
+        argument_presets:
+        - name: "id"
+          value:
+            from_session_variable: "x-hasura-user-id"
+        - name: "limit"
+          value:
+            static: 1
+    - name: User
+      fields:
+      - name: a
+      - name: b
+      - name: c
+```
+
+In the above example, if the client (as role "user") makes the query:
+
+```
+query {
+  user {
+    a
+    b
+  }
+}
+```
+
+then the query that is constructed for the remote schema is
+
+```
+query {
+  user(id: "<x-hasura-user-id>", limit: 1) {
+    a
+    b
+  }
+}
+```
+
+## Few questions to tackle:
+
+1. This will be a breaking change as all remote schemas will be hidden for all role except `admin`.
```

Comment:
I guess a nice UI would allow editing schema for a new role by either pre-populating with a completely unfiltered schema, or cloning an existing role's schema